### PR TITLE
Feature/vector spaces

### PR DESCRIPTION
**Pull request by Primetalk**

### CODE REVIEW

File: environment/src/main/scala/ru/primetalk/funtik/environment/geom2d/vectors.scala [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  def vectorSpaceForVector3d[Axis, Scalar](implicit cmodule: CModule[Axis, Scalar],field: Field[Scalar]): VectorSpace[Vector3d[Axis], Scalar] = new VectorSpace[Vector3d[Axis], Scalar] {
+    override implicit def scalar: Field[Scalar] = field
+    override def timesl(r: Scalar, v: Vector3d[Axis]): Vector3d[Axis] = Vector3d(cmodule.timesl(r, v.x), cmodule.timesl(r, v.y), cmodule.timesl(r, v.z))
+    override def negate(v: Vector3d[Axis]): Vector3d[Axis] = Vector3d(cmodule.negate(v.x), cmodule.negate(v.y), cmodule.negate(v.z))
+    override def zero: Vector3d[Axis] = Vector3d(cmodule.zero, cmodule.zero, cmodule.zero)
+    override def plus(a: Vector3d[Axis], b: Vector3d[Axis]): Vector3d[Axis] = Vector3d(cmodule.plus(a.x, b.x), cmodule.plus(a.y, b.y), cmodule.plus(a.z, b.z))
+  }
+
+
+//  implicit object fieldCmoduleInt extends Field[Int] with IntIsEuclideanRing with  CModule[Int, Int] {
```
Comment (sd1ver) [inline environment/src/main/scala/ru/primetalk/funtik/environment/geom2d/vectors.scala]: Do we need this?

File: environment/src/main/scala/ru/primetalk/funtik/environment/geom2d/vectorSpaces.scala [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  override def euclideanFunction(a: Int): BigInt = BigInt(a).abs
+  override def equot(a: Int, b: Int): Int        = spire.math.equot(a, b)
+  override def emod(a: Int, b: Int): Int         = spire.math.emod(a, b)
+}
+
+/** An implementation of 2d VectorSpace for pairs of Ints and Doubles.
+ * @tparam Axis is the type that is used to represent one of the coordinates.
+ * @tparam Scalar is either Int or Double that we use to scale vectors.
+ */
+class VectorSpaceProduct2[@sp(Int, Double) Axis, @sp(Int, Double) Scalar](implicit val scalar: Field[Scalar], val cmodule: CModule[Axis, Scalar])
```
Comment (sd1ver) [inline environment/src/main/scala/ru/primetalk/funtik/environment/geom2d/vectorSpaces.scala]: What are the difference between Vector2d and VectorSpaceProduct2?

File: environment/src/main/scala/ru/primetalk/funtik/environment/geom2d/vectorSpaces.scala [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  override def zero: (Axis, Axis) = (cmodule.zero, cmodule.zero)
+
+  override def plus(a: (Axis, Axis), b: (Axis, Axis)): (Axis, Axis) = (cmodule.plus(a._1, b._1), cmodule.plus(a._2, b._2))
+}
+
+/** An implementation of 3d VectorSpace for 3-tuples of Ints and Doubles.
+ * @tparam Axis is the type that is used to represent one of the coordinates.
+ * @tparam Scalar is either Int or Double that we use to scale vectors.
+ */
+class VectorSpaceProduct3[@sp(Int, Double) Axis, @sp(Int, Double) Scalar](implicit val scalar: Field[Scalar], val cmodule: CModule[Axis, Scalar])
```
Comment (sd1ver) [inline environment/src/main/scala/ru/primetalk/funtik/environment/geom2d/vectorSpaces.scala]: What are the difference between Vector3d and VectorSpaceProduct3?

File: environment/src/main/scala/ru/primetalk/funtik/environment/geom2d/vectors.scala [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+case class Vector2d[@sp(Int, Double) Axis](x: Axis, y: Axis) extends Vector[Axis, 2] {
+  def _1: Axis = x
+  def _2: Axis = y
+  def transpose: Vector2d[Axis] =
+    Vector2d(y, x)
+}
+
+case class Vector3d[@sp(Int, Double) Axis](x: Axis, y: Axis, z: Axis) extends Vector[Axis, 3] {
+  def _1: Axis = x
```
Comment (sd1ver) [inline environment/src/main/scala/ru/primetalk/funtik/environment/geom2d/vectors.scala:16]: Do we need this aliases?

Comment (Primetalk) [reply environment/src/main/scala/ru/primetalk/funtik/environment/geom2d/vectors.scala:16]: We may remove them as soon as we remove them in other places. At the moment they are being used in way too many places.

**Pull request merged**